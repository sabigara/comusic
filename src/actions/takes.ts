import { TakeState } from '../reducers/takes';
import { FileState } from '../reducers/files';
import { createAction, backendAPI } from '.';
import {
  fetchVerContentsSuccess,
  FETCH_VER_CONTENTS_SUCCESS,
} from './versions';

const RENAME_TAKE = 'RENAME_TAKE' as const;
const ADD_TAKE_REQUEST = 'ADD_TAKE_REQUEST' as const;
export const ADD_TAKE_SUCCESS = 'ADD_TAKE_SUCCESS' as const;
const ADD_TAKE_FAILURE = 'ADD_TAKE_FAILURE' as const;
const DELETE_TAKE_REQUEST = 'DELETE_TAKE_REQUEST' as const;
export const DELETE_TAKE_SUCCESS = 'DELETE_TAKE_SUCCESS' as const;
const DELETE_TAKE_FAILURE = 'DELETE_TAKE_FAILURE' as const;

export const ActionTypeName = {
  RENAME_TAKE,
  ADD_TAKE_SUCCESS,
  DELETE_TAKE_SUCCESS,
  // From outer modules.
  FETCH_VER_CONTENTS_SUCCESS,
};

export const renameTake = (id: string, name: string) => {
  return {
    type: RENAME_TAKE,
    id: id,
    payload: {
      name: name,
    },
  };
};

export const addTakeSuccess = (
  id: string,
  take: TakeState,
  file: FileState,
) => {
  return {
    type: ADD_TAKE_SUCCESS,
    id: id,
    payload: {
      take: take,
      file: file,
    },
  };
};

export const addTake = (trackId: string, formData: FormData) => {
  return async (dispatch: any) => {
    dispatch(createAction(ADD_TAKE_REQUEST, trackId));
    try {
      const resp = await backendAPI.addTake(trackId, formData);
      dispatch(addTakeSuccess(trackId, resp.take, resp.file));
    } catch (err) {
      dispatch(createAction(ADD_TAKE_FAILURE, trackId));
    }
  };
};

export const deleteTakeSuccess = (id: string) => {
  return {
    type: DELETE_TAKE_SUCCESS,
    id: id,
  };
};

export const deleteTake = (takeId: string) => {
  return async (dispatch: any) => {
    dispatch(createAction(DELETE_TAKE_REQUEST, takeId));
    try {
      await backendAPI.delTake(takeId);
      dispatch(deleteTakeSuccess(takeId));
    } catch (err) {
      dispatch(createAction(DELETE_TAKE_FAILURE, takeId));
    }
  };
};

export type ActionUnionType =
  | ReturnType<typeof renameTake>
  | ReturnType<typeof addTakeSuccess>
  | ReturnType<typeof deleteTakeSuccess>
  // From outer modules.
  | ReturnType<typeof fetchVerContentsSuccess>;
