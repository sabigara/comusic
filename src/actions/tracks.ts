/* eslint-disable @typescript-eslint/camelcase */
import { Track } from '../common/Domain';

const ADD_TRACK_REQUEST = 'ADD_TRACK_REQUEST' as const;
const ADD_TRACK_SUCCESS = 'ADD_TRACK_SUCCESS' as const;
const ADD_TRACK_FAILURE = 'ADD_TRACK_FAILURE' as const;
const DEL_TRACK_REQUEST = 'DEL_TRACK_REQUEST' as const;
const DEL_TRACK_SUCCESS = 'DEL_TRACK_SUCCESS' as const;
const DEL_TRACK_FAILURE = 'DEL_TRACK_FAILURE' as const;
const UPDATE_TRACK_PARAM = 'UPDATE_TRACK_PARAM' as const;
const CHANGE_ACTIVE_TAKE = 'CHANGE_ACTIVE_TAKE' as const;
const LOAD_ACTIVE_TAKE_REQUEST = 'LOAD_ACTIVE_TAKE_REQUEST' as const;
const LOAD_ACTIVE_TAKE_SUCCESS = 'LOAD_ACTIVE_TAKE_SUCCESS' as const;
const LOAD_ACTIVE_TAKE_FAILURE = 'LOAD_ACTIVE_TAKE_FAILURE' as const;

export const TrackActionTypeName = {
  UPDATE_TRACK_PARAM,
  CHANGE_ACTIVE_TAKE,
  LOAD_ACTIVE_TAKE_REQUEST,
  LOAD_ACTIVE_TAKE_SUCCESS,
  LOAD_ACTIVE_TAKE_FAILURE,
  ADD_TRACK_REQUEST,
  ADD_TRACK_SUCCESS,
  ADD_TRACK_FAILURE,
  DEL_TRACK_REQUEST,
  DEL_TRACK_SUCCESS,
  DEL_TRACK_FAILURE,
};

export const updateTrackParam = (
  trackId: string,
  param: string,
  value: number,
) => {
  return {
    type: UPDATE_TRACK_PARAM,
    id: trackId,
    payload: {
      param: param,
      value: value,
    },
  };
};

export const changeActiveTake = (trackId: string, activeTakeId: string) => {
  return {
    type: CHANGE_ACTIVE_TAKE,
    id: trackId,
    payload: {
      activeTakeId: activeTakeId,
    },
  };
};

export const loadActiveTakeRequest = (trackId: string) => {
  return {
    type: LOAD_ACTIVE_TAKE_REQUEST,
    id: trackId,
  };
};

export const loadActiveTakeSuccess = (trackId: string) => {
  return {
    type: LOAD_ACTIVE_TAKE_SUCCESS,
    id: trackId,
  };
};

export const loadActiveTakeFailure = (trackId: string, err: string) => {
  return {
    type: LOAD_ACTIVE_TAKE_FAILURE,
    id: trackId,
    err: err,
  };
};

export const addTrackSuccess = (verId: string, track: Track) => {
  return {
    type: ADD_TRACK_SUCCESS,
    id: verId,
    payload: {
      track: track,
    },
  };
};

export const delTrackSuccess = (trackId: string) => {
  return {
    type: DEL_TRACK_SUCCESS,
    id: trackId,
  };
};

export type TrackActionUnionType =
  | ReturnType<typeof updateTrackParam>
  | ReturnType<typeof changeActiveTake>
  | ReturnType<typeof loadActiveTakeRequest>
  | ReturnType<typeof loadActiveTakeSuccess>
  | ReturnType<typeof loadActiveTakeFailure>
  | ReturnType<typeof addTrackSuccess>
  | ReturnType<typeof delTrackSuccess>;
