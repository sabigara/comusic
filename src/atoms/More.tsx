import React from 'react';
import styled from 'styled-components';

const More: React.FC = () => {
  return (
    <Wrapper>
      <Dot />
      <Dot />
      <Dot />
    </Wrapper>
  );
};

const Wrapper = styled.div`
  height: 17px;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
`;

const Dot = styled.div`
  background-color: white;
  border-radius: 50%;
  width: 3px;
  height: 3px;
`;

export default More;
