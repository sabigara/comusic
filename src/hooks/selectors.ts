import { useSelector } from 'react-redux';

import { RootState } from '../reducers';

export const useTakes = (trackId: string) => {
  return useSelector((state: RootState) => {
    return state.takes.allIds
      .map((id) => state.takes.byId[id])
      .filter((take) => take.trackId === trackId);
  });
};

export const useActiveTakeId = (trackId: string) => {
  return useSelector((state: RootState) => {
    return state.tracks.byId[trackId].activeTake;
  });
};
