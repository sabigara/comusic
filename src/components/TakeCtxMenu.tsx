import React from 'react';
import { useDispatch } from 'react-redux';
import { ContextMenu, MenuItem, ContextMenuTrigger } from 'react-contextmenu';

import { deleteTake } from '../actions/takes';

type Props = {
  takeId: string;
};

type Data = {
  id: string;
  method: string;
};

const TakeContextMenu: React.FC<Props> = ({ takeId, children }) => {
  const dispatch = useDispatch();

  const handler = (e: any, data: Data) => {
    e.stopPropagation();
    switch (data.method) {
      case 'DELETE':
        dispatch(deleteTake(data.id));
        break;
    }
  };
  return (
    <div>
      <ContextMenu id={`take-ctx-${takeId}`}>
        <MenuItem data={{ method: 'DELETE', id: takeId }} onClick={handler}>
          Delete
        </MenuItem>
      </ContextMenu>
      <ContextMenuTrigger id={`take-ctx-${takeId}`} holdToDisplay={0}>
        {children}
      </ContextMenuTrigger>
    </div>
  );
};

export default TakeContextMenu;
