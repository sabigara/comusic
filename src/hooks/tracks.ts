import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { TrackParam, PlaybackStatus } from '../common/Domain';
import { RootState } from '../reducers';
import { ActionTypeName as ATN, createAction } from '../actions';
import {
  addTrackSuccess,
  delTrackSuccess,
  updateTrackParam,
  changeActiveTake,
  loadActiveTakeRequest,
  loadActiveTakeSuccess,
  loadActiveTakeFailure,
} from '../actions/tracks';
import useAudioAPI from './useAudioAPI';
import useBackendAPI from './useBackendAPI';
import useShouldTrackPlay from './useShouldTrackPlay';
import { useLoading } from './loading';

export const useAddTrack = (verId: string) => {
  const backendAPI = useBackendAPI();
  const audioAPI = useAudioAPI();
  const dispatch = useDispatch();

  return useCallback(async () => {
    dispatch(createAction(ATN.Track.ADD_TRACK_REQUEST, verId));
    try {
      const resp = await backendAPI.addTrack(verId);
      dispatch(addTrackSuccess(verId, resp));
      const trackAPI = audioAPI.loadTrack(resp.id);
      trackAPI.setVolume(resp.volume);
      trackAPI.setPan(resp.pan);
    } catch (err) {
      dispatch(
        createAction(ATN.Track.ADD_TRACK_FAILURE, verId, err.toString()),
      );
    }
  }, [verId]);
};

export const useDelTrack = () => {
  const backendAPI = useBackendAPI();
  const audioAPI = useAudioAPI();
  const dispatch = useDispatch();

  return useCallback(async (trackId: string) => {
    dispatch(createAction(ATN.Track.DEL_TRACK_REQUEST, trackId));
    try {
      await backendAPI.delTrack(trackId);
      dispatch(delTrackSuccess(trackId));
      audioAPI.getTrack(trackId)?.release();
    } catch (err) {
      err = err.toString();
      dispatch(createAction(ATN.Track.DEL_TRACK_FAILURE, trackId, err));
    }
  }, []);
};

export const useUpdateTrackParam = () => {
  const audioAPI = useAudioAPI();
  const dispatch = useDispatch();

  return useCallback(
    async (trackId: string, param: TrackParam, value: number) => {
      dispatch(updateTrackParam(trackId, param, value));
      const track = audioAPI.getTrack(trackId);
      switch (param) {
        case TrackParam.volume:
          track?.setVolume(value);
          break;
        case TrackParam.pan:
          track?.setPan(value);
          break;
        default:
          break;
      }
    },
    [],
  );
};

export const useSwitchMuteSolo = (trackId: string) => {
  const dispatch = useDispatch();
  const track = useSelector((state: RootState) => state.tracks.byId[trackId]);
  const shouldPlay = useShouldTrackPlay(trackId);

  const switchMute = useCallback(() => {
    dispatch(
      updateTrackParam(trackId, TrackParam.isMuted, Number(!track.isMuted)),
    );
  }, [trackId, track, shouldPlay]);

  const switchSolo = useCallback(() => {
    dispatch(
      updateTrackParam(trackId, TrackParam.isSoloed, Number(!track.isSoloed)),
    );
  }, [trackId, track, shouldPlay]);

  return [switchMute, switchSolo];
};

export const useRestartTrack = (trackId: string) => {
  const playback = useSelector(
    (state: RootState) => state.playback,
    (prev, current) => prev.status === current.status,
  );
  const audioAPI = useAudioAPI();
  const loadingTake = useLoading('LOAD_ACTIVE_TAKE');
  useEffect(() => {
    if (loadingTake) return;
    const trackAPI = audioAPI.tracks[trackId];
    if (!trackAPI) return;
    trackAPI.stop();
    if (playback.status === PlaybackStatus.Playing) {
      if (!trackAPI.isPlaying) {
        trackAPI.play(playback.time);
      }
    }
  }, [audioAPI.tracks, loadingTake, playback.status, playback.time, trackId]);
};

export const useLoadActiveTake = () => {
  const dispatch = useDispatch();
  const audioAPI = useAudioAPI();
  const fileById = useSelector((state: RootState) => state.files.byId);
  const takeById = useSelector((state: RootState) => state.takes.byId);

  return async (trackId: string, url?: string, takeId?: string) => {
    dispatch(loadActiveTakeRequest(trackId));
    try {
      if (!url) {
        if (!takeId) throw 'Must pass url or takeId';
        const take = takeById[takeId];
        if (!take) throw 'Provided takeId is not valid';
        const file = fileById[take.fileId];
        if (!file) throw 'Provided fileId is not ';
        url = file.url;
      }
      const trackAPI = audioAPI.getTrack(trackId);
      if (!trackAPI) throw 'No track to add take';
      trackAPI.stop();
      trackAPI.clearBuffer();
      await trackAPI.loadFile(url);
      dispatch(loadActiveTakeSuccess(trackId));
    } catch (err) {
      dispatch(loadActiveTakeFailure(trackId, err.toString()));
    }
  };
};

export const useChangeActiveTake = (trackId: string) => {
  const dispatch = useDispatch();
  const loadActiveTake = useLoadActiveTake();

  return useCallback(
    async (takeId) => {
      dispatch(changeActiveTake(trackId, takeId));
      loadActiveTake(trackId, undefined, takeId);
    },
    [loadActiveTake],
  );
};
