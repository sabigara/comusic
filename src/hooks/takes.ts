import { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { ActionTypeName as ATN, createAction } from '../actions';
import { addTakeSuccess, delTakeSuccess } from '../actions/takes';
import { AddTakeResp } from '../BackendAPI/interface';
import { useLoadActiveTake } from './tracks';
import useBackendAPI from './useBackendAPI';
import useAudioAPI from './useAudioAPI';
import { RootState } from '../reducers';

export const useAddTake = () => {
  const dispatch = useDispatch();
  const backendAPI = useBackendAPI();
  const loadActiveTake = useLoadActiveTake();

  return useCallback(
    async (trackId: string, formData: FormData) => {
      let resp: AddTakeResp | undefined = undefined;
      dispatch(createAction(ATN.Take.ADD_TAKE_REQUEST, trackId));
      try {
        resp = await backendAPI.addTake(trackId, formData);
        dispatch(addTakeSuccess(trackId, resp.take, resp.file));
      } catch (err) {
        dispatch(
          createAction(ATN.Take.ADD_TAKE_FAILURE, trackId, err.toString()),
        );
        return;
      }
      loadActiveTake(trackId);
    },
    [loadActiveTake],
  );
};

export const useDelTake = () => {
  const backendAPI = useBackendAPI();
  const dispatch = useDispatch();
  const audioAPI = useAudioAPI();
  const tracks = useSelector((state: RootState) => state.tracks.byId);
  const takes = useSelector((state: RootState) => state.takes.byId);

  return useCallback(
    async (takeId: string) => {
      dispatch(createAction(ATN.Take.DEL_TAKE_REQUEST, takeId));
      try {
        await backendAPI.delTake(takeId);
        dispatch(delTakeSuccess(takeId));
      } catch (err) {
        dispatch(
          createAction(ATN.Take.DEL_TAKE_FAILURE, takeId, err.toString()),
        );
        return;
      }
      const track = tracks[takes[takeId].trackId];
      const trackAPI = audioAPI.getTrack(track.id);
      if (track.activeTake === takeId) {
        trackAPI?.stop();
        trackAPI?.clearBuffer();
      }
    },
    [tracks, takes],
  );
};
