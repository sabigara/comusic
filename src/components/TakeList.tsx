import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import styled from 'styled-components';

import useAudioAPI from '../hooks/useAudioAPI';
import Color from '../common/Color';
import Label from '../atoms/Label';
import More from '../atoms/More';

import { changeActiveTake } from '../actions/trackList';

type Take = {
  id: string,
  name: string,
  fileURL: string,
}

type Props = {
  trackId: string,
}

const TakeList: React.FC<Props> = ({ trackId }) => {
  const state = useSelector((state: any) => {
    const track = state.trackList.filter((track: any) => track.id === trackId);
    return track ? track[0] : null
  },
    (prev, current) => prev.activeTakeId === current.activeTakeId
  );

  const [ mouseOver, setMouseOver] = useState(false);
  const [ mouseHoverId, setMouseHoverId ] = useState<string | null>(null);
  
  const dispatch = useDispatch();
  const audioAPI = useAudioAPI();

  const activeTakeURL = state.takeList.filter((take) => take.id === state.activeTakeId)[0].fileURL

  useEffect(() => {
    audioAPI.getTrack(trackId)!.loadFile(activeTakeURL);
  }, [audioAPI, trackId, activeTakeURL]);

  return (
    <Wrapper
      mouseOver={mouseOver}
      onMouseEnter={() => setMouseOver(true)}
      onMouseLeave={() => setMouseOver(false)}
    >
      {
        state.takeList.map((take, i) => {
          return (
            <TakeButton
              onMouseEnter={() => setMouseHoverId(take.id)}
              onMouseLeave={() => setMouseHoverId(null)}
              key={i}
              isActive={take.id === state.activeTakeId}
              onClick={(e) => {
                dispatch(changeActiveTake(trackId, take.id));
              }}
            >
              <ButtonLabel>{take.name}</ButtonLabel>
              {
                mouseHoverId === take.id
                ? <MoreWrapper onClick={(e) => e.stopPropagation()}><More/></MoreWrapper>
                : null
              }
            </TakeButton>
          );
        })
      }
      <label htmlFor="take_upload">
        <UploadButton>
         <UploadButtonLabel>+ Upload</UploadButtonLabel>
        </UploadButton>
      </label>
      <input
        type="file"
        id="take_upload"
        accept="audio/*"
        style={{display: 'none'}}
      />
    </Wrapper>
  );
}

const Wrapper = styled.div<{mouseOver: boolean}>`
  background-color: ${Color.Background};
  width: 150px;
  height: 170px;
  overflow-y: scroll;

  &::-webkit-scrollbar {
    background-color: ${Color.Background};
    width: 10px;
  }

  &::-webkit-scrollbar-thumb {
    background: #ccc;
    border-radius: 4px;
    width: 4px;
    display: ${props => props.mouseOver ? null : 'none'};
  }
`

const Button = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-radius: 3px;
  border: solid 1px ${Color.Border};
  box-sizing: border-box;
  width: 120px;
  height: 26px;
  margin: 7px auto;
  padding: 0 0 0 5px;
`

const TakeButton = styled(Button)<{isActive: boolean}>`
  background-color: ${props => {
    return props.isActive ? Color.Button.MuteOn : Color.Button.Disabled;
  }};
`

const UploadButton = styled(Button)`
  background-color: #0395EB;
  cursor: pointer;
`

const ButtonLabel = styled(Label)`
  font-size: 14px;
`

const UploadButtonLabel = styled(ButtonLabel)`
  cursor: pointer;
  font-size: 13px;
`

const MoreWrapper = styled.div`
  display: flex;
  justify-content: center;
  width: 14px;
  cursor: pointer;
`

export default TakeList;
