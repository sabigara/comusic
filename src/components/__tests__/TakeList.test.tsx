/* eslint-disable @typescript-eslint/no-explicit-any */
import * as React from 'react';
import * as Redux from 'react-redux';
import { shallow, mount } from 'enzyme';

import TakeList from '../TakeList';
import More from '../../atoms/More';
import TakeCtxMenu from '../TakeCtxMenu';
import * as TrackHooks from '../../hooks/tracks';
import * as TakeHooks from '../../hooks/takes';
import * as Selectors from '../../hooks/selectors';

jest.mock('../../AudioAPI/WebAudioAPI');
jest.mock('../../hooks/takes');

const mockTakes = [
  {
    id: '4c92e8b3-9456-4d8b-aec4-2d23b9bbaaff',
    name: 'with ride cymbal',
    trackId: '116c2e4e-ea34-49d7-8a2a-9b2da03d7048',
    file: '16229d8b-e4e3-41e6-85e3-5ce847ae7fa4',
  },
  {
    id: '7599b9e5-6055-4917-b183-e59bd4ec429e',
    name: 'with hihat',
    trackId: '116c2e4e-ea34-49d7-8a2a-9b2da03d7048',
    file: '21c64940-fbca-470d-9b5d-99db5c98efee',
  },
];

function mockHooks() {
  const mockUseTakes = jest
    .spyOn(Selectors, 'useTakes')
    .mockReturnValue(mockTakes as any);
  const mockActiveTakeId = jest
    .spyOn(Selectors, 'useActiveTakeId')
    .mockReturnValue('4c92e8b3-9456-4d8b-aec4-2d23b9bbaaff');
  const mockUseAddTake = jest.spyOn(TakeHooks, 'useAddTake');
  const mockUseDispatch = jest.spyOn(Redux, 'useDispatch');
  const mockUseState = jest.spyOn(React, 'useState');
  const mockUseCallback = jest.spyOn(React, 'useCallback');
  // Just pass through the given function.
  mockUseCallback.mockImplementation((fn: any) => fn);
  const mockChangeActiveTake = jest.fn();
  jest
    .spyOn(TrackHooks, 'useChangeActiveTake')
    .mockReturnValue(mockChangeActiveTake);
  return {
    mockUseState,
    mockUseCallback,
    mockUseDispatch,
    mockUseTakes,
    mockUseAddTake,
    mockActiveTakeId,
    mockChangeActiveTake,
  };
}

function shallowRender() {
  return shallow(<TakeList trackId={'116c2e4e-ea34-49d7-8a2a-9b2da03d7048'} />);
}

describe('TakeList', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    mockHooks();
    const container = shallowRender();
    expect(container).toMatchSnapshot();
  });

  it('has take-button named "with ride cymbal activated"', () => {
    mockHooks();
    const container = shallowRender();
    const takeBtn = container.find('.take-button').first();
    expect(takeBtn.text()).toBe('with ride cymbal');
    expect(takeBtn.prop('isActive')).toBe(true);
  });

  it('calls changeActiveTake with proper takeId', () => {
    const { mockChangeActiveTake } = mockHooks();
    const container = shallowRender();
    const takeBtn1 = container.find('.take-button').at(0);
    const takeBtn2 = container.find('.take-button').at(1);
    // Shouldn't be called to prevent unnecessary update.
    takeBtn1.simulate('click');
    expect(mockChangeActiveTake.mock.calls.length).toBe(0);
    takeBtn2.simulate('click');
    expect(mockChangeActiveTake.mock.calls[0][0]).toBe(
      '7599b9e5-6055-4917-b183-e59bd4ec429e',
    );
  });

  it('prints More component on mouse enter', () => {
    mockHooks();
    const container = shallowRender();
    const takeBtn = container.find('.take-button').first();
    takeBtn.simulate('mouseenter');
    expect(container).toMatchSnapshot();
  });

  it('shows more icon on mouse enter and hide on leave', () => {
    mockHooks();
    const container = shallowRender();
    const takeBtn = container.find('.take-button').first();
    takeBtn.simulate('mouseenter');
    // takeBtn.contains(<More />) should be true, but it doesn't
    // contain any under TackCtxMenu. So instead assert by container.
    expect(container.contains(<More />)).toBe(true);
    takeBtn.simulate('mouseleave');
    expect(takeBtn.contains(<More />)).toBe(false);
  });

  it('calls addTake() with proper args', () => {
    const { mockUseAddTake } = mockHooks();
    const mockAddTake = jest.fn();
    mockUseAddTake.mockReturnValue(mockAddTake);
    const container = shallowRender();
    const input = container.find('input').first();
    input.simulate('change', {
      target: {
        files: [{ name: 'dummy.wav' }],
      },
    });
    expect(mockAddTake.mock.calls.length).toBe(1);
    expect(mockAddTake.mock.calls[0][0]).toBe(
      '116c2e4e-ea34-49d7-8a2a-9b2da03d7048',
    );
    const body = mockAddTake.mock.calls[0][1];
    expect(body.get('name')).toBe('dummy.wav');
  });
});
