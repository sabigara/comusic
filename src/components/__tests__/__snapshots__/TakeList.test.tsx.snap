// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TakeList prints More component on mouse enter 1`] = `
<styled.div
  mouseOver={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Styled(styled.div)
    className="take-button"
    isActive={true}
    key="4c92e8b3-9456-4d8b-aec4-2d23b9bbaaff"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <Styled(styled.span)>
      with ride cymbal
    </Styled(styled.span)>
    <TakeContextMenu
      takeId="4c92e8b3-9456-4d8b-aec4-2d23b9bbaaff"
    >
      <styled.div
        onClick={[Function]}
      >
        <More />
      </styled.div>
    </TakeContextMenu>
  </Styled(styled.div)>
  <Styled(styled.div)
    className="take-button"
    isActive={false}
    key="7599b9e5-6055-4917-b183-e59bd4ec429e"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <Styled(styled.span)>
      with hihat
    </Styled(styled.span)>
  </Styled(styled.div)>
  <label
    htmlFor="take-upload-116c2e4e-ea34-49d7-8a2a-9b2da03d7048"
  >
    <Styled(styled.div)>
      <Styled(Styled(styled.span))>
        + Upload
      </Styled(Styled(styled.span))>
    </Styled(styled.div)>
  </label>
  <input
    accept="audio/*"
    id="take-upload-116c2e4e-ea34-49d7-8a2a-9b2da03d7048"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
</styled.div>
`;

exports[`TakeList should match the snapshot 1`] = `
<styled.div
  mouseOver={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Styled(styled.div)
    className="take-button"
    isActive={true}
    key="4c92e8b3-9456-4d8b-aec4-2d23b9bbaaff"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <Styled(styled.span)>
      with ride cymbal
    </Styled(styled.span)>
  </Styled(styled.div)>
  <Styled(styled.div)
    className="take-button"
    isActive={false}
    key="7599b9e5-6055-4917-b183-e59bd4ec429e"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <Styled(styled.span)>
      with hihat
    </Styled(styled.span)>
  </Styled(styled.div)>
  <label
    htmlFor="take-upload-116c2e4e-ea34-49d7-8a2a-9b2da03d7048"
  >
    <Styled(styled.div)>
      <Styled(Styled(styled.span))>
        + Upload
      </Styled(Styled(styled.span))>
    </Styled(styled.div)>
  </label>
  <input
    accept="audio/*"
    id="take-upload-116c2e4e-ea34-49d7-8a2a-9b2da03d7048"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
</styled.div>
`;
